openedx/core/djangoapps/cache_toolbox/tests/test_middleware.py

```diff
@@ -45,3 +49,22 @@ def test_session_change_cms(self):
         home_url = reverse('home')
         # Studio login redirects to LMS login
         self._test_change_session_hash(home_url, settings.LOGIN_URL + '?next=' + home_url, target_status_code=302)
+
+    def test_user_logout_on_session_hash_change(self):
+        """
+        Verify that if a user's session auth hash and the request's hash
+        differ, the user is logged out:
+         - session is flushed
+         - request user is changed to Anonymous user
+         - logged in cookies are marked for deletion
+        """
+        session_id = self.client.session.session_key
+        safe_cookie_data = SafeCookieData.create(session_id, self.user.id)
+        self.request.COOKIES[settings.SESSION_COOKIE_NAME] = str(safe_cookie_data)
+        SafeSessionMiddleware().process_request(self.request)   # preparing session
+
+        with patch.object(User, 'get_session_auth_hash', return_value='abc123'):
+            CacheBackedAuthenticationMiddleware().process_request(self.request)
+            assert self.request.session.get(SESSION_KEY) is None
+            assert self.request.user == AnonymousUser()
+            assert self.request.need_to_delete_cookie
```

Comment:
When does the actual deletion happen? Is it possible to assert on the `settings.SESSION_COOKIE_NAME` no longer existing? If not, and it needs to be this partial check, please assert that `need_to_delete_cookie` does not yet exist above.

Reply:
The actual deletion happens in the `process_response` function of `SafeSessionMiddleware` so its not possible to assert on it here but I have added an assert above that `need_to_delete_cookie` does not exist before.

Reply:
Why can’t the test also call process_response?

Reply:
@robrap it's possible and I can update the test to check `SafeSessionMiddleware` response but there is a test that already exists that checks this functionality in `SafeSessionMiddleware` tests. (ref: https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/safe_sessions/tests/test_middleware.py#L231-L234)

We can replicate that test and test this here as well but I guess there is no need for code duplication. But I will leave it to you. I can update the PR if you think we should test it here too.

Reply:
* Although you are setting `settings.SESSION_COOKIE_NAME` in the test, it might be good to add an assertion that the cookie is on the request beforehand, and assert afterwards that it is not on the request.
* I would remove any reference to `need_to_delete_cookie`, which is just an implementation detail, but the main point of the test is that the cookie gets removed. I think it is fine to have this more end-to-end test, rather than testing the implementation details that come in-between.

Reply:
@robrap I have updated the test, I hope it makes sense now.